plugins/woo-ai/src/product-short-description/product-short-description-button-container.tsx

```diff
@@ -0,0 +1,133 @@
+import { useTinyEditor } from '../hooks';
+import { getPostId, recordTracksFactory } from '../utils';
+import { translateApiErrors as getApiError } from '../utils/apiErrors';
+
+const recordShortDescriptionTracks = recordTracksFactory(
+	'short_description_completion',
+	() => ( {
+		post_id: getPostId(),
+	} )
+);
+
+export function WriteShortDescriptionButtonContainer() {
+	const { createWarningNotice } = useDispatch( 'core/notices' );
+
+	const [ fetching, setFetching ] = useState< boolean >( false );
+	const tinyEditor = useTinyEditor();
+	const shortTinyEditor = useTinyEditor( 'excerpt' );
+	const [ postContent, setPostContent ] = useState( '' );
+
+	const handleUseCompletionError = ( err: UseCompletionError ) => {
+		createWarningNotice( getApiError( err.code ?? '' ) );
+		setFetching( false );
+		// eslint-disable-next-line no-console
+		console.error( err );
+	};
+
+	const { requestCompletion, completionActive, stopCompletion } =
+		useCompletion( {
+			feature: WOO_AI_PLUGIN_FEATURE_NAME,
+			onStreamMessage: ( content ) => {
+				shortTinyEditor.setContent( content );
+			},
+			onStreamError: handleUseCompletionError,
+			onCompletionFinished: ( reason, content ) => {
+				recordShortDescriptionTracks( 'stop', {
+					reason,
+					character_count: content.length,
+				} );
+
+				setFetching( false );
+			},
+		} );
+
+	useEffect( () => {
+		recordShortDescriptionTracks( 'view_button' );
+	}, [] );
+
+	// This effect sets up the 'init' listener to update the 'isEditorReady' state
+	useEffect( () => {
+		const editor = tinyEditor.getEditorObject();
+		if ( editor ) {
+			// Set the content on initial page load.
+			setPostContent( tinyEditor.getContent() );
+			// Register a listener for the tinyMCE editor to update the postContent state.
+			const changeHandler = () => {
+				setPostContent( tinyEditor.getContent() );
+			};
+			editor.on( 'input', changeHandler );
+			editor.on( 'change', changeHandler );
+			return () => {
+				editor.off( 'input', changeHandler );
+				editor.off( 'change', changeHandler );
+			};
+		}
+	}, [ tinyEditor ] );
+
+	const writeItForMeEnabled =
+		! fetching && postContent.length >= MIN_DESC_LENGTH_FOR_SHORT_DESC;
+
+	const buildPrompt = (): string => {
+		return [
+			'Please write a high-converting Meta Description for the WooCommerce product description below.',
+			'It should strictly adhere to the following guidelines:',
+			'It should entice someone from a search results page to click on the product link.',
+			'It should be no more than 155 characters so that the entire meta description fits within the space provided by the search engine result without being cut off or truncated.',
+			'It should explain what users will see if they click on the product page link.',
+			'Do not wrap in double quotes or use any other special characters.',
+			'It should include the target keyword for the product.',
+			`Here is the full product description: \n${ tinyEditor.getContent() }`,
```

Comment:
Have you tried sending it more data (categories, attributes, title, etc.) to see if it improves the result? Should we send it the AI Customization settings we recently added (tone and store description)?

Reply:
I think Joel's assumption here (and mine as well) was that GPT could infer the data it needs from the description. But, it's certainly worth trying. I'll let you know how it goes.

Reply:
Adding the tone of voice does work!:
![Screenshot 2023-08-31 at 13 04 46](https://github.com/woocommerce/woocommerce/assets/6723003/d2aec8cb-1d74-4598-b46c-241d0b0f9af6)

Unfortunately, it seems like adding more to the prompt results in it forgetting previous instructions, namely, that we do not want the short description to be surrounded by quotes. This happened probably 50% of the time when I tested. :(

![Screenshot 2023-08-31 at 13 04 33](https://github.com/woocommerce/woocommerce/assets/6723003/83a1fb76-44a9-4fc7-adae-ed0b0f757592)

![Screenshot 2023-08-31 at 13 05 01](https://github.com/woocommerce/woocommerce/assets/6723003/bb85dabb-6603-442c-aa88-95c331041371)

I think we'll want a couple of things before we can confidently write prompts like these:

- GPT 4 or equivalent (LLama2?)
- A way to rigorously test our prompts against test cases

I think it would be great to spend a short sprint cycle doing some spikes on these two points.

For now, I'd like to leave the prompt as-is, but, I have refactored it into a function because it was duplicated.

https://github.com/woocommerce/woocommerce/pull/39805/commits/dbc72b0f7436fc6c83c65ad72885beff540b607a